source/user/mcscf.rst

```diff
@@ -1,57 +1,299 @@
-or CI coefficients (not that useful unless doing a DMRG-CASSCF calculation) or
-both.  For example:
+Minimal examples for each type of calculations are shown below.
 
-.. literalinclude:: /../examples/mcscf/13-restart.py
+CASCI
+"""""
+.. code-block:: python
 
+  import pyscf
+  mol = pyscf.M(
+      atom = 'O 0 0 0; O 0 0 1.2',
+      basis = 'ccpvdz',
+      spin = 2)
+  myhf = mol.RHF().run()
+  ncas, nelecas = (6,8)
+  mycas = myhf.CASCI(ncas, nelecas).run()
 
 
- The third argument for CASCI/CASSCF is the size of CAS space; the
-fourth argument is the number of electrons.  By default, the CAS
-solver determines the alpha-electron number and beta-electron number
-based on the attribute :attr:`Mole.spin`.  In the above example, the
-number of alpha electrons is equal to the number of beta electrons,
-since the ``mol`` object is initialized with ``spin=0``.  The spin
-multiplicity of the CASSCF/CASCI solver can be changed by the fourth
-argument::
+CASSCF
+""""""
+.. code-block:: python
 
-  >>> mc = mcscf.CASSCF(m, 4, (4,2))
-  >>> print('E(CASSCF) = %.9g' % mc.kernel()[0])
-  E(CASSCF) = -149.609461
-  >>> print('S^2 = %.7f, 2S+1 = %.7f' % mcscf.spin_square(mc))
-  S^2 = 2.0000000, 2S+1 = 3.0000000
+  import pyscf
+  mol = pyscf.M(
+      atom = 'O 0 0 0; O 0 0 1.2',
+      basis = 'ccpvdz',
+      spin = 2)
+  myhf = mol.RHF().run()
+  ncas, nelecas = (6,8)
+  mycas = myhf.CASSCF(ncas, nelecas).run()
 
-The two integers in the tuple represent the number of alpha and beta electrons.
-Although it is a triplet state, the solution might not be correct since the
-CASSCF is based on the incorrect singlet HF ground state.  Starting from the
-ROHF ground state, we have::
+Like many other modules in PySCF, :mod:`mcscf` works with density-fitting (:numref:`user_df`) and x2c (:numref:`user_x2c`).
+It can also be used in its unrestricted formulation, please see :source:`examples/mcscf/60-uhf_based_ucasscf.py` for an example.
+An important feature of :mod:`mcscf` is that it can interface with external CI solver such as DMRG, FCIQMC, or selected CI methods, see the external projects for more details.
 
-  >>> mc = mcscf.CASSCF(rhf3, 4, 6)
-  >>> print('E(CASSCF) = %.9g' % mc.kernel()[0])
-  E(CASSCF) = -149.646746
 
-The energy is lower than the RHF initial guess.
-.. We can also use the UHF ground
-.. state to start a CASSCF calculation::
-.. 
-..   >>> mc = mcscf.CASSCF(uhf3, 4, 6)
-..   >>> print('E(CASSCF) = %.9g' % mc.kernel()[0])
-..   E(CASSCF) = -149.661324
-..   >>> print('S^2 = %.7f, 2S+1 = %.7f' % mcscf.spin_square(mc))
-..   S^2 = 3.9713105, 2S+1 = 4.1091656
-.. 
-.. Woo, the total energy is even lower.  But the spin is contaminated.
+
+Picking an Active Space
+-----------------------
+In general, selecting an active space can be cumbersome and PySCF offers severals ways to facilitate this process.
+There are four strategies and all require that you specify the number of electrons and orbitals in the active space.
```

Comment:
Maybe it is good to explain that PySCF does a clever thing under the hood. The program actually uses the molecule charge to count electrons and add orbitals to active space from occupied or empty HF orbitals according to the input.

I also think it is important to distinguish the choice of the active space itself, which is fundamental from choice of starting orbitals that will be corrected by the SCF procedure.

Reply:
1) Yes. However, this is pretty obvious and is also used in other programs (at least Q-Chem does it). Also, this is how active spaces are always defined in the literature: (ne,nmo).

2) I don't agree: one cannot distinguish between the active space and the choice of the active orbitals. The orbitals **define** your active space. (E.g. if you have a diatomic molecule, the symmetry of the orbitals already defines your CAS problem: say, the 5s5p active space  = 10 sigma 5 pi, and this defines the Hamiltonian. You can always find the orbitals that minimize the CAS energy for this choice of the active space.)

3) "initial guess orbitals" is as correct as "starting orbitals". CAS-SCF is like SCF i.e. Hartree-Fock theory. You have an initial guess for the orbitals, and then you optimize them by minimizing the CAS-SCF energy.

Reply:
>if you have a diatomic molecule, the symmetry of the orbitals already defines your CAS problem

The fact that the symmetry of inactive, active and virtual spaces is defined by starting orbitals is another 'smart' thing in PySCF. This actually should probably discussed as well.

In Molcas and Molpro to define CAS one has to give number of inactive and active orbitals per irrep together with number of active electrons. The common *(ne, nmo)* is just a short hand notation that implies a list of *nmo* per irrep.